### PR TITLE
builtins: implement ST_Collect aggregate

### LINKED ISSUE #48903: geo/geomfn: implement ST_Collect({geometry})
Implement `ST_Collect` on arguments {geometry}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_Collect.html).

*Observers: Please react to this issue if you need this functionality.*

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

The following additional guidance has been issued on implementing this function:

Extend the aggregate builtins to include collect.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-08-30T17:39:33Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by erikgrinaker**
Also contains `geomfn.Collect()` for the binary version of `ST_Collect` (#48902), but the builtin could not be implemented since it cannot have the same name as an aggregate function. The code is left behind for when this constraint is lifted, but I'm happy to remove it entirely if preferred.

The aggregation optimistically uses a multitype, but once it encounters an incompatible geometry it converts the existing multitype into a collection, using additional memory and CPU time - ideas for how to avoid this are welcome.

Test coverage is less than great, but I could not see any existing convention for comprehensive testing of aggregate functions.

Release justification: low risk, high benefit changes to existing
functionality

Release note (sql change): Implement the geometry aggregate builtin
ST_Collect.

Closes #48903.

### CODE REVIEW

**Comment by erikgrinaker** [comment]
CC @otan

**Comment by otan** [comment]
Thanks! You'll need to bump the distsql version too -- see https://github.com/cockroachdb/cockroach/pull/53127/files and the `pkg/sql/execinfra` changes

File: pkg/sql/opt/operator.go [hunk 1]
```diff
@@ -396,7 +397,7 @@ func AggregatesCanMerge(inner, outer Operator) bool {
 
 	case AnyNotNullAggOp, BitAndAggOp, BitOrAggOp, BoolAndOp,
 		BoolOrOp, ConstAggOp, ConstNotNullAggOp, FirstAggOp,
-		MaxOp, MinOp, STMakeLineOp, STExtentOp, STUnionOp, SumOp, SumIntOp, XorAggOp:
+		MaxOp, MinOp, STMakeLineOp, STCollectOp, STExtentOp, STUnionOp, SumOp, SumIntOp, XorAggOp:
```
Comment (otan) [inline pkg/sql/opt/operator.go:400]: i think it shouldn't be able to merge because the ordering of st_collect can change.

**Review by otan — COMMENTED** [review]
nice!

**Comment by erikgrinaker** [comment]
> You'll need to bump the distsql version too

Done, thanks. Will this affect the release justification?

File: pkg/sql/opt/operator.go (hunk 1 above)
Comment (erikgrinaker) [reply pkg/sql/opt/operator.go]: Good point, thanks - updated.

**Comment by otan** [comment]
> Done, thanks. Will this affect the release justification?

nah should be fine!

**Review by rytaft — APPROVED** [review]
opt changes :lgtm: 

> Reviewed 8 of 10 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53645#-:-MG4aHS44y1FqEkZB73F:b-yql68r)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @otan)

___
*[pkg/geo/geomfn/collections.go, line 24 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/53645#-MG4SoWh7sp_f6WXA-jC:-MG4SoWh7sp_f6WXA-jD:buu4kmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8658854bb08e5dfbf0d5715eefd165c122863911/pkg/geo/geomfn/collections.go#L24)):*
> ```Go
> // This is the binary version of `ST_Collect()`, but since it's not possible to have an aggregate
> // and non-aggregate function with the same name (different args), this is not used. Code is left
> // behind for when we add support for this. Be sure to handle NULL args when adding the builtin for
> ```

[nit] comments should be 80 columns wide

**Comment by erikgrinaker** [comment]
Fixed comment width, rebased onto `master`, and updated tests. Should be good to go.

**Review by otan — APPROVED** [review]

**Pull request merged**